prow/jenkins/jenkins.go

```diff
@@ -401,14 +425,121 @@ func getJobName(spec *prowapi.ProwJobSpec) string {
+	jenkinsPath := fmt.Sprintf("/job/%s/api/json", jenkinsJobName)
+
+	return jenkinsPath
+}
+
+// getBuildPath builds a path to trigger a regular build for this job
+func getBuildPath(spec *prowapi.ProwJobSpec) string {
+	jenkinsJobName := getJobName(spec)
+	jenkinsPath := fmt.Sprintf("/job/%s/build", jenkinsJobName)
+
+	return jenkinsPath
+}
+
+// getBuildWithParametersPath builds a path to trigger a build with parameters for this job
+func getBuildWithParametersPath(spec *prowapi.ProwJobSpec) string {
 	jenkinsJobName := getJobName(spec)
 	jenkinsPath := fmt.Sprintf("/job/%s/buildWithParameters", jenkinsJobName)
 
 	return jenkinsPath
 }
 
+// GetJobInfo retrieves Jenkins job information
+func (c *Client) GetJobInfo(spec *prowapi.ProwJobSpec) (*JobInfo, error) {
+	path := getJobInfoPath(spec)
+	c.logger.Debugf("getJobInfoPath: %s", path)
+
+	data, err := c.Get(path)
+
+	if err != nil {
+		c.logger.Errorf("Failed to get job info: %v", err)
+		return nil, err
+	}
+
+	var jobInfo JobInfo
+
+	if err := json.Unmarshal(data, &jobInfo); err != nil {
+		return nil, fmt.Errorf("Cannot unmarshal job info from API: %v", err)
+	}
+
+	c.logger.Tracef("JobInfo: %+v", jobInfo)
+
+	return &jobInfo, nil
+}
+
+// JobParameterized tells us if the Jenkins job for this ProwJob is parameterized
+func (c *Client) JobParameterized(jobInfo *JobInfo) bool {
+	for _, prop := range jobInfo.Property {
+		if prop.ParameterDefinitions != nil && len(prop.ParameterDefinitions) > 0 {
+			return true
+		}
+	}
+
+	return false
+}
+
+// EnsureBuildableJob attempts to detect a job that hasn't yet ran and populated
+// its parameters. If detected, it tries to run a build until the job parameters
+// are processed, then it aborts the build.
+func (c *Client) EnsureBuildableJob(spec *prowapi.ProwJobSpec) error {
+	jobInfo, err := c.GetJobInfo(spec)
+
+	if err != nil {
+		return err
+	}
+
+	isParameterized := c.JobParameterized(jobInfo)
+
+	c.logger.Tracef("JobHasParameters: %v", isParameterized)
+
+	if isParameterized || len(jobInfo.Builds) > 0 || err != nil {
+		return err
+	}
+
+	path := getBuildPath(spec)
+	c.logger.Debugf("getBuildPath: %s", path)
+
+	resp, err := c.request(http.MethodPost, path, nil, true)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 201 {
+		return fmt.Errorf("EnsureBuildableJob build job response not 201: %s", resp.Status)
+	}
+
+	backoff := wait.Backoff{
+		Duration: time.Duration(5) * time.Second,
+		Factor:   1,
+		Jitter:   1,
+		Steps:    10,
+	}
+
+	return wait.ExponentialBackoff(backoff, func() (bool, error) {
+		c.logger.Debugf("Waiting for job %v to become parameterized", spec.Job)
+
+		jobInfo, _ := c.GetJobInfo(spec)
+		isParameterized := false
+
+		if jobInfo != nil {
+			isParameterized = c.JobParameterized(jobInfo)
+
+			if isParameterized && jobInfo.LastBuild != nil {
+				c.logger.Debugf("Job %v is now parameterized, aborting the build", spec.Job)
+				c.Abort(getJobName(spec), &jobInfo.LastBuild)
```

Comment:
does this ignore the error? should at least log it

Reply:
I figured it's not worth failing the whole thing if for some reason the previous build wasn't aborted, certainly not against logging the error.